### PR TITLE
Collection - Add two entries:

### PR DESCRIPTION
**Pull request by phweyland**
Collection - Add two entries:
- module
- order

if plugins/lighttable/collect/module=TRUE in darktablerc.

I've missed the way to get a more user friendly module name. Any idea ?

image  image  image

What does surprise me is to have so many custom images. Most of them come from dt 2.6 if I'm not mistaken.

### CODE REVIEW

**Comment by TurboGit** [comment]
Nice !

One question, do you even sleep or are you a robot :)

I'll review ASAP !

**Comment by Nilvus** [comment]
@TurboGit: as @phweyland said, it would be good to see the user module name and not the file name. For example, having tone contrast instead of atrous.

**Comment by TurboGit** [comment]
> I've missed the way to get a more user friendly module name. Any idea ?

Would this routine in imagiop.h do the trick:

`gchar *dt_iop_get_localized_name(const gchar *op);`

**Comment by TurboGit** [comment]
@Nilvus : Yes, was on it and I think I found the solution.

**Comment by TurboGit** [comment]
> What does surprise me is to have so many custom images. Most of them come from dt 2.6 if I'm not mistaken.

Or non edited pictures ?

File: src/libs/collect.c [hunk 1]
```diff
@@ -1508,6 +1508,40 @@ static void list_view(dt_lib_collect_rule_t *dr)
...
 
+      case DT_COLLECTION_PROP_MODULE: // module
+        snprintf(query, sizeof(query),
+                 "SELECT operation, 1, COUNT(*) AS count"
+                 " FROM main.images AS mi"
+                 " JOIN "
+                 "   (SELECT imgid, operation FROM main.history) h"
+                 "  ON h.imgid = mi.id"
+                 " WHERE %s"
+                 " GROUP BY operation"
```
Comment (TurboGit) [inline src/libs/collect.c]: I'm wondering if group_by and order_by won't be slow without an index for history.operation. Have you seen speed issue ?

**Comment by phweyland** [comment]
> Would this routine in imagiop.h do the trick:
> 
> `gchar *dt_iop_get_localized_name(const gchar *op);`

I'll look at this.

> Or non edited pictures ?

I'm not sure. I've migrated from lr early 2018 and a lot of raw files have been updated through ligthroom.c.
However I've imported also a lot of a lot of lr exports (jpg) in order not to loose too much work. :-)
 
The other doubt is that I try to keep my dt 2.7 database. I'll make try with a true 2.6 database.

File: src/libs/collect.c (hunk 1 above)
Comment (phweyland) [reply src/libs/collect.c]: I haven't seen any issue there (so far).

Comment (TurboGit) [reply src/libs/collect.c]: So nothing to do, will see if needed later.

**Comment by phweyland** [comment]
module collection lists now the localized names.
I still see some issues:
- filmic selection gives filmic + filmic rgb images
- very often the number of displayed images is lower than the counted ones in modules list. Should be the same. I don't know yet the cause (may be the state of my db).

**Comment by phweyland** [comment]
image

**Comment by phweyland** [comment]
Except the readiness of darktable.iop at start time that seems OK for me.
 
Doubt. In history there the module numeric field which seems to identify the module too. It could faster to use it instead of operation. However this number is not available on dt_iop_module_so_t.

File: src/common/collection.c [hunk 2]
```diff
@@ -1502,6 +1502,27 @@ static gchar *get_query_string(const dt_collection_properties_t property, const
       query = dt_util_dstrcat(query, "(id %s group_id)", (strcmp(escaped_text, _("group leaders")) == 0) ? "=" : "!=");
       break;
 
+    case DT_COLLECTION_PROP_MODULE: // dev module
+      {
+        dt_iop_set_module_name_table();
```
Comment (TurboGit) [inline src/common/collection.c]: We want to call this only once at startup I think. So better place it during initialization of the application in darktable.c.

Comment (TurboGit) [reply src/common/collection.c]: I would propose at the end of dt_init() routine at least after the call to   dt_iop_load_modules_so();

Comment (phweyland) [reply src/common/collection.c]: It's what I've done. But this doesn't solve that the list is not ready when collect.c refreshes the screen.

Comment (phweyland) [reply src/common/collection.c]: sorry, not collect.c but dt_collection_update()

Comment (phweyland) [reply src/common/collection.c]: dt_collection_update() is called by dt_collection_new() and dt_selection_new()
around line 900 in darktable.c

Comment (TurboGit) [reply src/common/collection.c]: So let's try to move both calls around line 980 after dt_iop_load_modules_so(); It should work.

Comment (phweyland) [reply src/common/collection.c]: seg fault ...
I've tried to call dt_collection_update(darktable.collection) after updating the list.
I get this:
image
Like a kind of redraw was missing ...

Comment (phweyland) [reply src/common/collection.c]: but   dt_control_queue_redraw_center(); doesn't help

**Comment by phweyland** [comment]
I think that's ok now.

**Comment by phweyland** [comment]
Migrating from a 2.6 db, all the modules are custom for me.
I'd expected all legacy, wouldn't I ?

**Comment by phweyland** [comment]
> Migrating from a 2.6 db, all the modules are custom for me.

If I'm not mistaken discrepancy comes between mask_manager (legacy order) and denoiseprofile (images) for all images...

**Comment by TurboGit** [comment]
> If I'm not mistaken discrepancy comes between mask_manager (legacy order) and denoiseprofile (images) for all images...

Could be the case indeed, the mask_manager may move in the iop list IIRC. Not something that will break things but indeed annoying ! Probably something to fix in this area.

**Comment by phweyland** [comment]
Hummm. 2.6 has not been released. So it's already an in between db.
But I've tried with a db I've marked a 2.4.4 (29/9/2018) and the oldest db I have (10/4/2018) with the same result.

**Comment by phweyland** [comment]
Is the legacy_order list really correct ? If I remember correctly, lut3d was not born 

> . 2.6 has not been released.

Sorry, it has been actually.

**Comment by phweyland** [comment]
> 2.6 has not been released. 

Sorry, my mistake,  it has been released actually.

File: src/libs/collect.c [hunk 3]
```diff
@@ -2169,6 +2204,8 @@ void gui_init(dt_lib_module_t *self)
   GtkBox *box;
   GtkWidget *w;
 
+  const gboolean show_module = dt_conf_get_bool("plugins/lighttable/collect/module");
```
Comment (TurboGit) [inline src/libs/collect.c]: Probably time to remove this now ? This was for debug purpose, right ? Applying your PR I was surprised to not find the two new filters !

File: src/libs/collect.h [hunk 4]
```diff
@@ -36,7 +36,9 @@ const char *dt_lib_collect_string[] = { N_("film roll"),    N_("folders"),     N
                                         N_("rights"),       N_("lens"),        N_("focal length"),
                                         N_("ISO"),          N_("aperture"),    N_("exposure"),
                                         N_("aspect ratio"), N_("filename"),    N_("geotagging"),
-                                        N_("grouping"),     N_("local copy") };
+                                        N_("grouping"),     N_("local copy"),
+                                        // last two ones - skipped if not configured in darktablerc
```
Comment (TurboGit) [inline src/libs/collect.h]: And the comment here.

File: src/libs/collect.c (hunk 3 above)
Comment (phweyland) [reply src/libs/collect.c]: My idea was not to bother people who are not concerned with migration with this.
But of course we can show them all the time. not problem.

Comment (phweyland) [reply src/libs/collect.c]: > Activate those filter by default.

Ok, I do this.

**Comment by phweyland** [comment]
> Maybe just ignoring the placement of the mask_manager in the list would be sufficient.

What about if I replace 0 in module_order by 1 everywhere ?

**Comment by TurboGit** [comment]
> What about if I replace 0 in module_order by 1 everywhere ?

You'll break your edit, what you can try is to:
- replace 0 by 1 everywhere
- remove the iop_list by NULL for those entries

And backup your DB :)

**Comment by phweyland** [comment]
> what you can try is to:
> 
>     * replace 0 by 1 everywhere
> 
>     * remove the iop_list by NULL for those entries

Seems to work fine.
I can set v3.0 (I get 2 in version, iop_list still NULL) or change the order (I get 0 in version and the custom list in iop_list).

**Comment by TurboGit** [comment]
Well probably a bad advice. This will work for images without multi instance but it will break the ones that are really custom.

**Comment by TurboGit** [comment]
So you should do this only for image without multi insurance. So only for images without a multi priority > 0 in the history.

**Comment by phweyland** [comment]
Understood.
As my 2.7 backups are too messy I will have to consolidate on my last 2.6.
If you plan to mitigate this, I'll way. If not I may change the code as you suggested above to get the legacy where it should be.

**Comment by phweyland** [comment]
Skipping mask_manager in dt_ioppr_get_iop_order_list_kind() for legacy gives me 100% of legacy.
So I think that at this time I was not using multi instances ... :)

**Comment by phweyland** [comment]
With the change you suggested above, I've still 100% legacy from 2.6.
with these discrepancies:

legacy order - image
tonemap bilateral (most of them)
denoiseprofile bilateral
tonemap rotatepixels
tonemap bilateral
rawprepare collect

**Comment by phweyland** [comment]
Checking the multi instances, ... would not there be an issue in darktable.c line 1293 ?

            if(next
               && strcmp(((dt_iop_order_entry_t *)(l->data))->operation,
                         ((dt_iop_order_entry_t *)(next->data))->operation))

I would replace by:

            if(next
               && !strcmp(((dt_iop_order_entry_t *)(l->data))->operation,
                         ((dt_iop_order_entry_t *)(next->data))->operation))

**Comment by phweyland** [comment]
When I import my 2.6 db, I have some images without iop_order (not in the table module_order).
So I've added an order entry "none". 

When I open one of these images in darkroom the displayed order is v3.0. This image is then updated as v3.0 in module_order while I don't make any change on it.

> do you even sleep or are you a robot :)

Don't worry for the time, I'm on Brasilia time zone for the time being :)

**Comment by TurboGit** [comment]
> Checking the multi instances, ... would not there be an issue in darktable.c line 1293 ?
> 
> ```
>         if(next
>            && strcmp(((dt_iop_order_entry_t *)(l->data))->operation,
>                      ((dt_iop_order_entry_t *)(next->data))->operation))
> ```
> 
> I would replace by:
> 
> ```
>         if(next
>            && !strcmp(((dt_iop_order_entry_t *)(l->data))->operation,
>                      ((dt_iop_order_entry_t *)(next->data))->operation))
> ```

Yes :( Again this stupid strcmp() mistake ! I hate C :(

**Comment by phweyland** [comment]
My turn to worry about your sleep time... :)

**Comment by TurboGit** [comment]
But this was not an issue, it was only a way to avoid writing the iop_list in module_order. With the wrong check we have many iop_list written even though the order is a canonical one (legacy or v3.0). So not big deal which is luck!

**Comment by phweyland** [comment]
I've slightly modified the code to keep multiples instance as custom and all others set as legacy (as I suppose that order difference in 2.6 was only an error).

I guess that images without iop_order were missing in the previous history. I don't know if this case is a normal one, but the query in line database.c lines 1200-1204 could retrieve also the those records and the corresponding images set with a safe iop_order.

**Comment by phweyland** [comment]
>all others set as legacy (as I suppose that order difference in 2.6 was only an error).

And BTW I'll save myself from the tons of warning I get with the wrong iop order messages :)

**Comment by TurboGit** [comment]
> and all others set as legacy

My idea was to keep non edited picture to have them using v3.0 order first time entering in darkroom.

**Comment by TurboGit** [comment]
Anyway, merged now. Thanks.

**Comment by phweyland** [comment]
> y idea was to keep non edited picture to have them using v3.0 order first time entering in darkroom.

Are non edited pictures the ones which are not in the history table ?
If yes, that's the current behavior and that's ok.

**Comment by Nilvus** [comment]
@phweyland: Thanks for this add, really useful. Anyway, I'm quite surprised to see that both activated and deactivated modules are proposed on module filter. Would it be possible to only have activated module on this filter ? For example, I have some old images where I see a module I don't use anymore but I use on the past because it just remains on history. One use I see actually to this filter is to see if I forgot some images with an old module I want to remove is always there. Not possible by now.

**Comment by TurboGit** [comment]
@Nilvus : Indeed, makes more sense to have only activated modules I suppose.

**Comment by phweyland** [comment]
Agreed. I'll do that.
I didn't bother with that because I usually clean up my history as much as I can. But this is boring, I'll try to see if I can create an option to make the history clean-up complete :)

**Comment by phweyland** [comment]
BTW do you think it is worth to have the option to hide module and order entry ? 

Other question. I think we could remove the history entry as it is covered by order entry.
history - non altered = order - none ( I could change this one as 'non altered' to appear normal).

**Comment by phweyland** [comment]
Is the entry 'order' well named. If not what is your preference ? history (replacing the existing one) ? module order ?

**Comment by johnny-bit** [comment]
@phweyland - I tried small fix in #4301 - can you check if you have noticable difference in timings between query `WHERE h.enabled=1 and m.name LIKE '%s'`  and `WHERE m.name LIKE '%s' and h.enabled=1`?
My tests indicate that checking for `LIKE` first is a bit faster, but my library is tiny (~couple thousands photos and a bit more history items)

**Comment by TurboGit** [comment]
> BTW do you think it is worth to have the option to hide module and order entry ?

No,, I think we can remove this.

**Comment by Nilvus** [comment]
> Agreed. I'll do that.
> I didn't bother with that because I usually clean up my history as much as I can. But this is boring, I'll try to see if I can create an option to make the history clean-up complete :)

Why not an option to make history clean-up complete.

**Comment by TurboGit** [comment]
My preference will go for "module order". This is the name of the lib to change module order, so seems good for consistency.

**Comment by Nilvus** [comment]
> Other question. I think we could remove the history entry as it is covered by order entry.
> history - non altered = order - none ( I could change this one as 'non altered' to appear normal).

We could remove this but with an option button to do that. Some times, it's good to see which modules we disable.

**Comment by phweyland** [comment]
> > Other question. I think we could remove the history entry as it is covered by order entry.
> > history - non altered = order - none ( I could change this one as 'non altered' to appear normal).
> 
> We could remove this but with an option button to do that. Some times, it's good to see which modules we disable.

If I understand your wish, it is not related to 'history' entry but to 'module' where we could have a configuration option to include disabled ones.
But we could also have two entries per module, one for enabled and disabled .

**Comment by phweyland** [comment]
> can you check if you have noticable difference in timings between query `WHERE h.enabled=1 and m.name LIKE '%s'` and `WHERE m.name LIKE '%s' and h.enabled=1`?

I'll try to measure that.

**Comment by Nilvus** [comment]
> > > Other question. I think we could remove the history entry as it is covered by order entry.
> > > history - non altered = order - none ( I could change this one as 'non altered' to appear normal).
> > 
> > 
> > We could remove this but with an option button to do that. Some times, it's good to see which modules we disable.
> 
> If I understand your wish, it is not related to 'history' entry but to 'module' where we could have a configuration option to include disabled ones.
> But we could also have two entries per module, one for enabled and disabled .

I have no wish there as I let history as it is. My only wish related to this PR is what what @johnny-bit do on his #4301 PR. Actual history is great as it is for me.

But for those wanted to clean their history, I think the best is just to add a button to delete unused entries on history (so disabled modules). This way, that let other users to keep history as it is, so also with entries about disabled modules.

**Comment by phweyland** [comment]
> > can you check if you have noticable difference in timings between query `WHERE h.enabled=1 and m.name LIKE '%s'` and `WHERE m.name LIKE '%s' and h.enabled=1`?
> 
> I'll try to measure that.

I don't see any difference 0,07 average in both cases with 50.000 images

**Comment by johnny-bit** [comment]
> But for those wanted to clean their history, I think the best is just to add a button to delete unused entries on history (so disabled modules)

I thought that's what "compress history" did... Turns out it's a bit more complicated...

> I don't see any difference 0,07 average in both cases with 50.000 images

Thanks! So I'll think I'll do as @TurboGit advised and do `enabled` first, then `name LIKE` part.

**Comment by Nilvus** [comment]
> > But for those wanted to clean their history, I think the best is just to add a button to delete unused entries on history (so disabled modules)
> 
> I thought that's what "compress history" did... Turns out it's a bit more complicated...

Indeed not, compress history, just compressed modules changed more than once a time in the image (so it's simplify the history by merging similar entries.

Anyway, I have not ask to change history, that's just a suggestion from @phweyland and I have just answer to this. As nobody else ask for that, maybe the simple way is to not touch to history. I really like the way it is.

**Pull request merged**